src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,256 @@
+            "Append",
+            "Cast",
+            "Distinct",
+            "GroupBy",
+            "OfType",
+            "OrderBy",
+            "OrderByDescending",
+            "Prepend",
+            "Reverse",
+            "Select",
+            "SelectMany",
+            "Skip",
+            "SkipLast",
+            "SkipWhile",
+            "Take",
+            "TakeLast",
+            "TakeWhile",
+            "ThenBy",
+            "ThenByDescending",
+            "Where");
+
+        /// <summary>
+        /// Linq methods deferring its first two parameters to be enumerated.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_linqTwoParametersDeferredMethods = ImmutableArray.Create(
+            "Concat",
+            "Except",
+            "GroupJoin",
+            "Intersect",
+            "Join",
+            "Union");
+
+        internal abstract GlobalFlowStateDictionaryFlowOperationVisitor CreateOperationVisitor(
+            GlobalFlowStateDictionaryAnalysisContext context,
+            WellKnownSymbolsInfo wellKnownSymbolsInfo,
+            IMethodSymbol? getEnumeratorMethod);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(CompilationStartAction);
+        }
+
+        private void CompilationStartAction(CompilationStartAnalysisContext context)
+        {
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+        }
+
+        private void OnOperationBlockStart(OperationBlockStartAnalysisContext operationBlockStartAnalysisContext)
+        {
+            var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(operationBlockStartAnalysisContext.Compilation);
+
+            var oneParameterDeferredMethods = GetOneParameterDeferredMethods(wellKnownTypeProvider);
+            var twoParametersDeferredMethods = GetTwoParametersDeferredMethods(wellKnownTypeProvider);
+
+            var oneParameterEnumeratedMethods = GetOneParameterEnumeratedMethods(wellKnownTypeProvider);
+            var twoParametersEnumeratedMethods = GetTwoParametersEnumeratedMethods(wellKnownTypeProvider);
+
+            var compilation = operationBlockStartAnalysisContext.Compilation;
+            var additionalDeferTypes = GetTypes(compilation, s_additionalDeferredTypes);
+
+            var wellKnownSymbolsInfo = new WellKnownSymbolsInfo(
+                oneParameterDeferredMethods,
+                twoParametersDeferredMethods,
+                oneParameterEnumeratedMethods,
+                twoParametersEnumeratedMethods,
+                additionalDeferTypes);
+
+            var potentialDiagnosticOperationsBuilder = ImmutableHashSet.CreateBuilder<IOperation>();
```

Comment:
Use `PooledHashSet<IOperation>.GetInstance()` here, and free the builder inside operation end action callback.